src/gui/widgets/typed_debugger.cc

```diff
@@ -0,0 +1,967 @@
+            const std::regex data_regex(R"(([\w\s\*\[\]]+),(\w+),(\d+))");
+            std::smatch matches;
+            if (std::regex_match(s, matches, data_regex)) {
+                data.type = matches[1].str();
+                data.name = matches[2].str();
+                data.size = std::stoul(matches[3].str());
+            }
+
+            switch (importType) {
+                case ImportType::DataTypes:
+                    fields.push_back(data);
+                    break;
+                case ImportType::Functions:
+                    func.arguments.push_back(data);
+
+                    ++linePartsCounter;
+                    // Address + name + 4 arguments = 6.
+                    if (linePartsCounter == 6) {
+                        break;
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+        switch (importType) {
+            case ImportType::DataTypes:
+                m_structs[name] = fields;
+                m_typeNames.push_back(name);
+                break;
+            case ImportType::Functions:
+                m_functions[address] = func;
+                break;
+            default:
+                break;
+        }
+    }
+
+    if (importType == ImportType::DataTypes) {
+        std::sort(m_typeNames.begin(), m_typeNames.end(), [](std::string left, std::string right) {
+            for (auto& letter : left) {
+                letter = tolower(letter);
+            }
+            for (auto& letter : right) {
+                letter = tolower(letter);
+            }
+            return left.compare(right) < 0;
+        });
+    }
+}
+
+PCSX::Widgets::TypedDebugger::TypedDebugger(bool& show) : m_show(show), m_listener(g_system->m_eventBus) {
+    m_listener.listen<PCSX::Events::ExecutionFlow::Reset>([this](const auto& event) {
+        m_displayedWatchData.clear();
+        for (const auto* bp : m_watchBreakpoints) {
+            g_emulator->m_debug->removeBreakpoint(bp);
+        }
+        m_watchBreakpoints.clear();
+        m_disabledInstructions.clear();
+
+        m_displayedFunctionData.clear();
+        for (const auto* bp : m_functionBreakpoints) {
+            g_emulator->m_debug->removeBreakpoint(bp);
+        }
+        m_functionBreakpoints.clear();
+        m_disabledFunctions.clear();
+    });
+
+    m_listener.listen<PCSX::Events::ExecutionFlow::SaveStateLoaded>([this](const auto& event) {
+        // When loading a savestate, ensure instructions and functions that have been toggled off are disabled, just in
+        // case they weren't when the savestate was created.
+        uint8_t* const memData = g_emulator->m_mem->m_psxM;
+        constexpr uint32_t memBase = 0x80000000;
+
+        for (const auto& disabledInstruction : m_disabledInstructions) {
+            const auto instructionAddress = disabledInstruction.first;
+            constexpr uint8_t nop[4] = {0x00, 0x00, 0x00, 0x00};
```

Comment:
You can use `mips/common/util/encoder.hh` for these, if you want :) Might be a bit more clear.